Detectors/EMCAL/workflow/src/EventBuilderSpec.cxx

```diff
@@ -0,0 +1,140 @@
+    auto celldata = ctx.inputs().getByPos(posCells, islot);
+    auto subspecification = framework::DataRefUtils::getHeader<header::DataHeader*>(celldata)->subSpecification;
+    // Discard message if it is a deadbeaf message (empty timeframe)
+    if (subspecification == 0xDEADBEEF) {
+      continue;
+    }
+    cells[subspecification] = ctx.inputs().get<gsl::span<o2::emcal::Cell>>(celldata);
+  }
+  for (decltype(numSlotsTriggerRecords) islot = 0; islot < numSlotsTriggerRecords; islot++) {
+    auto trgrecorddata = ctx.inputs().getByPos(posTriggerRecords, islot);
+    auto subspecification = framework::DataRefUtils::getHeader<header::DataHeader*>(trgrecorddata)->subSpecification;
+    // Discard message if it is a deadbeaf message (empty timeframe)
+    if (subspecification == 0xDEADBEEF) {
+      continue;
+    }
+    triggers[subspecification] = ctx.inputs().get<gsl::span<o2::emcal::TriggerRecord>>(trgrecorddata);
+  }
+
+  for (auto interaction : connectRangesFromSubtimeframes(triggers)) {
+    eventCells.clear();
+    for (auto [subspec, triggerrecord] : interaction.mRangesSubtimeframe) {
+      LOG(DEBUG) << interaction.mInteractionRecord.bc << " / " << interaction.mInteractionRecord.orbit << ": Receiving " << triggerrecord.getEntries() << " digits for subspec " << subspec;
+      auto cellcont = cells.find(subspec);
+      if (cellcont != cells.end()) {
+        for (auto& cell : gsl::span<const Cell>(cellcont->second.data() + triggerrecord.getFirstEntry(), triggerrecord.getEntries())) {
+          eventCells.emplace_back(cell);
+        }
+      } else {
+        LOG(ERROR) << "No cell container found for subspec " << subspec;
+      }
+    }
+    int start = outputCells.size();
+    if (eventCells.size()) {
+      std::sort(eventCells.begin(), eventCells.end(), [](const Cell& lhs, const Cell& rhs) { return lhs.getTower() < rhs.getTower(); });
+      std::copy(eventCells.begin(), eventCells.end(), std::back_inserter(outputCells));
+    }
+    outputTriggers.emplace_back(interaction.mInteractionRecord, interaction.mTriggerType, start, eventCells.size());
+  }
+
+  ctx.outputs().snapshot(framework::Output{dataorigin, "CELLS", 0, framework::Lifetime::Timeframe}, outputCells);
+  ctx.outputs().snapshot(framework::Output{dataorigin, "CELLSTRGR", 0, framework::Lifetime::Timeframe}, outputTriggers);
+}
+
+std::set<EventBuilderSpec::RangeCollection> EventBuilderSpec::connectRangesFromSubtimeframes(const std::unordered_map<o2::header::DataHeader::SubSpecificationType, gsl::span<const TriggerRecord>>& triggerrecords) const
+{
+  std::set<RangeCollection> events;
+
+  // Search interaction records from all subevents
+  std::set<o2::InteractionRecord> allInteractions;
+  for (auto& [subspecification, trgrec] : triggerrecords) {
+    for (auto rec : trgrec) {
+      auto eventIR = rec.getBCData();
+      if (allInteractions.find(eventIR) == allInteractions.end())
+        allInteractions.insert(eventIR);
```

Comment:
Why do you need `find`?  `insert` will anyway call it, like this you are doing it twice.

Reply:
The goal of these lines was to avoid double counting.  The loop is doing a set merge of the sets of trigger records assuming the sets could differ (empty timeframe in either of the sets could be an option). In this case the interaction record should only be added in case not already present.

Reply:
std::set by definition contains only unique elements, if the trig.record is already there, insert will do nothing.

Reply:
OK, then I will remove the check.